xdsl/transforms/lower_riscv_func.py

```diff
@@ -85,14 +85,16 @@ def match_and_rewrite(self, op: riscv_func.FuncOp, rewriter: PatternRewriter):
 
 class InsertExitSyscallOp(RewritePattern):
     @op_type_rewrite_pattern
-    def match_and_rewrite(self, op: riscv_func.FuncOp, rewriter: PatternRewriter):
-        if op.sym_name.data != "main":
+    def match_and_rewrite(self, op: riscv_func.ReturnOp, rewriter: PatternRewriter):
+        parent_op = op.parent_op()
+        if (
+            not isinstance(parent_op, riscv_func.FuncOp)
+            or parent_op.sym_name.data != "main"
```

Comment:
I don't think checking for parent op is "main" is enough.

Reply:
Care to elaborate? MLIR does not allow early returns, therefore this should be enough, right? Can you provide a counterexample?

Reply:
I'm not sure what you mean @AntonLydike

Reply:
I was replying to @kingiler, your comment just ninja'd it way in-between, I'll edit it to be more clearer!

Reply:
Ah yes I see, it's about return being possibly nested. Today we don't have the HasParent trait but we probably should eventually, guaranteeing that it's in the right place.